src/relay/backend/contrib/codegen_c/codegen.cc

```diff
@@ -290,8 +290,37 @@ runtime::Module CCompiler(const ObjectRef& ref) {
   return csource.CreateCSourceModule(ref);
 }
 
+/*!
+ * \brief A visitor to add the constants used as params for MetadataModule.
+ */
+struct CCompilerConstantUpdater : public ExprVisitor {
```

Comment:
why is this needed?

Reply:
Essentially just so that I can write a test case, because Ethos-N registers a 'null' constant updater. This also has the benefit of removing the implicit coupling between the CCompiler and the default ConstantUpdater (currently they must both adopt the same naming and visiting convention).

Reply:
I think this doesn't couple to the updater though because it does nothing as what the constant updater does.

Reply:
The concern is that the ConstantUpdater has no direct coupling with the external codegens, and yet it is generic. That is to say, the C Compiler codegen _must_ visit and name its constants in the same way that the ConstantUpdater chooses to, or you get a runtime failure. However, this relationship isn't enforced by any mechanism and relies on the codegen implementer to understand they must follow the convention. A simple example where this may happen is if a codegen only wants to save some constants and not others - the Ethos-N codegen being an extreme example where it wants to save none of them.

Reply:
How about just add a sentence in the file description to explain that the constant updater directly uses the default/generic one so that we don't really need to have the exactly same one here? Since C compiler is not really for production use, I think we should keep is simple and clean.

Reply:
@zhiics , What about the other way ? We make every codegen register a constant updater -- as you say if most of the codegen are based on JSONRuntime maybe that can implement the generic one and which could be overridden if the codegen requires. This way it creates a nice division of responsibilities -- constant creation and consumption is at more or less the same set of sources.

Reply:
@manupa-arm I think one of the problems we are trying to avoid is adding more and more explicit registries to the external codegens. But it doesn't seems necessary to me to register a constant updater for each codegen because currently probably only ethos-n needs an empty updater. I think we can just fallback to generic one if there is no need for a special updater.

Reply:
I do agree we need to avoid more and more explicit registries but this registry is more of a requirement to enable codegens based on other languages to be linked to TVM -- the same argument as we do relay.ext.<codegen_name> registry.

The problem I see with the current design of constant updating in this manner is if you look at the VisitExpr of ConstantNode, the correct lookup of constant is entirely dependent of the constant_idx that is generated by a generic constant updater living outside the codegen. This locks the down the design of external codegens to traverse in the same way GenericConstant updater is traversing and also blocks any modification to constant if such is required by the external codegen (e.g., compression, adding a header, etc.).

To highlight the further the issue of locking down the traversal will become more damaging when the graph becomes less-linear (the ones that could be generated via NASs in future). In which case we might have a degree of freedom of scheduling (i.e. we might want to schedule one operator before the other for various reasons -- such as caching, memory, performance that depends on the ext backend). 

Although, I appreciate that its makes the life quite easy of a tvm developer to have the generic constant updater but however I think it should live in the codegen itself. It also improves readabilty as well -- because now we can figure out the why the constants are indexed in the same way and how it retrieves the exact constant due to the fact the codegen post it to the params in the first place -- rather than having a generic constant updater living outside the codegen sources.

Maybe its outside of the scope for this PR -- but this PR exposed a symptom of the real problem, IMHO.

Reply:
If you agree to what I say, I think we should slowly start showing this style of design in ccompiler and promote it to be used by external codegen developers to make it future proof and deprecate the generic one living outside the codegen once we have every codegen comply to it.

Reply:
@manupa-arm Thanks for clarification. I am not arguing the necessity of constant updater for codegens. My point is that we can probably just add to the ones that needs it. To keep consistency, you can also register it for each of the other codegens but we don't really need to copy the code everywhere if there can fully reuse the generic one.

Reply:
So the reason I added it for ccompiler is really just so that I could write a test for this code pathway. The Ethos-N updater doesn't really do anything so it's not great for testing. I could just temporarily register a new constant updater for ccompiler in the test instead, would that be better?

Reply:
Yeah, that would be better. Thanks.